flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/BarrierHandlerAction.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io.checkpointing;
+
+import org.apache.flink.runtime.checkpoint.CheckpointException;
+import org.apache.flink.runtime.checkpoint.channel.InputChannelInfo;
+import org.apache.flink.runtime.io.network.api.CheckpointBarrier;
+
+import java.io.IOException;
+
+/**
+ * Represents a state in a state machine of processing a checkpoint. There are 4 base states:
+ *
+ * <ul>
+ *   <li>Waiting for an aligned barrier
+ *   <li>Collecting aligned barriers
+ *   <li>Waiting for an unaligned barrier
+ *   <li>Collecting unaligned barriers
+ * </ul>
+ *
+ * <p>Additionally depending on the configuration we can switch between aligned and unaligned
+ * actions.
+ */
+interface BarrierHandlerAction {
```

Comment:
If you really want to do the state machine route (which is probably a good idea), how about actually calling it `BarrierHandlerState`. I had a hard time initially to recognize the state machine.
I'd name the internal state then `ChannelState`.

Reply:
Initially I named it `BarrierHandlerState` however in a discussion with @pnowojski we said it's a bit weird that a state does not actually have a state (the `BarrierHandlerState/Action` does not hold any state apart from the `ChannelState`.

But I actually like your naming scheme better. If @pnowojski is fine with it, I'll rename to `BarrierHandlerState` and `ChannelState`.

Reply:
I find it perfectly normal for a state machine state to not contain any state. In fact, I find it strange otherwise. For me, a state machine is ideally encoded into an `enum` and we explicitly pass a state object to each method.
If a state machine state contains state, I think we have not gained much in terms of making state transitions apparent.